crates/aptos/src/move_tool/bytecode.rs

```diff
@@ -0,0 +1,308 @@
+    },
+};
+use anyhow::Context;
+use async_trait::async_trait;
+use clap::Parser;
+use itertools::Itertools;
+use move_binary_format::{
+    binary_views::BinaryIndexedView, file_format::CompiledScript, CompiledModule,
+};
+use move_bytecode_source_map::{mapping::SourceMapping, utils::source_map_from_file};
+use move_command_line_common::files::{
+    MOVE_COMPILED_EXTENSION, MOVE_EXTENSION, SOURCE_MAP_EXTENSION,
+};
+use move_coverage::coverage_map::CoverageMap;
+use move_disassembler::disassembler::{Disassembler, DisassemblerOptions};
+use move_ir_types::location::Spanned;
+use std::{
+    fs,
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+const REVELA_EXE_ENV: &str = "REVELA_EXE";
+#[cfg(target_os = "windows")]
+const REVELA_EXE: &str = "revela.exe";
+#[cfg(not(target_os = "windows"))]
+const REVELA_EXE: &str = "revela";
+
+const DISASSEMBLER_EXTENSION: &str = "mv.asm";
+const DECOMPILER_EXTENSION: &str = "mv.move";
+
+/// Disassemble the Move bytecode pointed to in the textual representation
+/// of Move bytecode.
+///
+/// For example, if you want to disassemble an on-chain package `PackName` at account `0x42`:
+/// 1. Download the package with `aptos move download --account 0x42 --package PackName --bytecode`
+/// 2. Disassemble the package bytecode with `aptos disassemble --package-path PackName/bytecode_modules`
+#[derive(Debug, Parser)]
+pub struct Disassemble {
+    #[clap(flatten)]
+    pub command: BytecodeCommand,
+}
+
+/// Decompile the Move bytecode pointed to into Move source code.
+///
+/// For example, if you want to decompile an on-chain package `PackName` at account `0x42`:
+/// 1. Download the package with `aptos move download --account 0x42 --package PackName --bytecode`
+/// 2. Decompile the package bytecode with `aptos decompile --package-path PackName/bytecode_modules`
+#[derive(Debug, Parser)]
+pub struct Decompile {
+    #[clap(flatten)]
+    pub command: BytecodeCommand,
+}
+
+#[derive(Debug, Clone, Copy, Default)]
+enum BytecodeCommandType {
+    // Need to set a default for technical reasons, but we override it anyway below.
+    #[default]
+    Disassemble,
+    Decompile,
+}
+
+#[derive(Debug, Parser)]
+pub struct BytecodeCommand {
+    /// Set programmatically
+    #[clap(skip)]
+    command_type: BytecodeCommandType,
+
+    /// Treat input file as a script (default is to treat file as a module)
+    #[clap(long)]
+    pub is_script: bool,
+
+    /// The path to a directory containing Move bytecode files with the extension `.mv`.
+    /// The tool will process all files find in this directory
+    ///
+    /// If present, a source map at the same location ending in `.mvsm` and the source
+    /// file itself ending in`.move` will be processed by the tool.
+    #[clap(long)]
+    pub package_path: Option<PathBuf>,
+
+    /// Alternatively to a package path, path to a single bytecode file which should be processed.
+    #[clap(long)]
+    pub bytecode_path: Option<PathBuf>,
```

Comment:
Nit: It is generally more idiomatic to check for mutually exclusive args at the clap level, that way you get consistent error messages. See https://stackoverflow.com/a/76315811/3846032.

Reply:
Done. Good to know this.